File: arrays_Pancake_sort.java
import java.io.*;
import java.util.*; 

class Solution {

  static int[] pancakeSort(int[] arr) {

    
    // we are gradually reducing range
    for(int length = arr.length; length >=0; length--) {
      
      int biggestValue = arr[0];
      int biggestIndex = 0;
      
      // looking for biggest in current range
      for( int i = 0; i < length; i++){ // this range will be reduced (length)
        if(arr[i] > biggestValue) {
          biggestValue = arr[i];
          biggestIndex = i;
        }
      }
      arr = flip(arr,biggestIndex); // we put BIGGEST in front
      arr = flip(arr, length-1 ); // we need it to don't tuch already flipped previous BIGGEST ONES
      
    }
    
    return arr;
    
  }
  
  // flip function flips order in given range (first k elems) of array
    static int[] flip(int [] arr , int k ){
      int endIndex = k;
      for(int i = 0 ; i <=  k ; i ++, endIndex--){

        if( endIndex != i && endIndex > i ) {
          int swap = arr[i];
          arr[i] = arr[endIndex];
          arr[endIndex] = swap;
        } else {
          break;
        }

      }
      return arr;
    }

  public static void main(String[] args) {

  }

}

File: arrays_Pancake_sort.js
function pancakeSort(arr){

	for (var length = arr.length; length >= 0; length--) {
		var biggestValue = arr[0];
		var biggestIndex = 0;
		// looking for biggest in the current range
		for (var i = 0; i < length; i++) {
			if(arr[i] > biggestValue){
				biggestValue = arr[i];
				biggestIndex = i;
			}
		}
		// we put BIGGEST in front
		arr = flip(arr, biggestIndex);
		// 
		arr = flip(arr, length-1);
	}
	return arr;
}

	// flip function
	function flip(arr, k){
		var endIndex = k;

		for (var i = 0; i <= k; i++ ) {
			if( endIndex != i && endIndex >i ){
				var swap = arr[i];
				arr[i] = arr[endIndex];
				arr[endIndex] = swap;
			} else {
				break;
			}
			endIndex--
		}
	}

	console.log( pancakeSort([1, 5, 4, 3, 2]) );

